### PR TITLE
Share a feature extractor between classifiers

### PR DESCRIPTION
**Pull request by vinvinod**
This is an attempt to share same feature extractor between classifiers.
Modified only named_entity_extractor and text_categorizer.

### CODE REVIEW

**Comment by davisking** [comment]
Cool, thanks for the PR.  I'll take a look at it this week.  

On first glance though, I see there is a mixture of tabs and spaces in the change.  Can you update it so it's only spaces, this way the code looks uniform and the github change view isn't messed up?

**Comment by vinvinod** [comment]
I have converted all tabs to spaces.

**Comment by davisking** [comment]
I still see a bunch of tabs and bad indenting. :(

**Comment by vinvinod** [comment]
I am really sorry! I should have been thorough.
Hopefully, now I haven't missed anything.

**Comment by davisking** [comment]
No worries.

File: mitielib/include/mitie.h [hunk 1]
```diff
@@ -151,6 +151,13 @@ extern "C"
             - The returned object MUST BE FREED by a call to mitie_free().
             - If the object can't be created then this function returns NULL.
     !*/
+    typedef struct mitie_total_word_feature_extractor mitie_total_word_feature_extractor;
+    MITIE_EXPORT mitie_named_entity_extractor* mitie_load_named_entity_extractor_pure_model_without_feature_extractor (
+        const char* filename
+    );
+    /*!
+	Added for Rasa_nlu multitenancy
```
Comment (davisking) [inline mitielib/include/mitie.h]: What does this mean?

More generally, all the functions in this file should be properly documented.  e.g. what does this function do?  What are the requirements for calling it?

File: examples/python/text_categorizer_multiple_models.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+trainer2.add_labeled_text(tokenize("Finish"),"negative")
+trainer2.add_labeled_text(tokenize("Close"),"negative")
+trainer2.num_threads = 4
+cat2 = trainer2.train()
+cat2.save_to_disk("new_text_categorizer_2_pure_model.dat",pure_model=True)
+
+
+text_feature_extractor = total_word_feature_extractor(fe_filename)
+
+cat1_new = text_categorizer("new_text_categorizer_1_pure_model.dat", None, True)
```
Comment (davisking) [inline examples/python/text_categorizer_multiple_models.py]: The user shouldn't have to supply None,True as an argument.  The code can figure out that the file on disk is a pure model so it should not require the user to supply this redundant information.

File: mitielib/include/mitie/total_word_feature_extractor.h [hunk 3]
```diff
@@ -128,7 +128,7 @@ namespace mitie
                 feats.set_size(0);
                 return;
             }
-
+            dlib::matrix<float,0,1> temp;
```
Comment (davisking) [inline mitielib/include/mitie/total_word_feature_extractor.h]: Why is this here?

File: mitielib/include/mitie/word_morphology_feature_extractor.h [hunk 4]
```diff
@@ -64,6 +64,7 @@ namespace mitie
                 - #feats.size() == get_num_dimensions()
         !*/
         {
+            std::vector<dlib::uint16> hits;
```
Comment (davisking) [inline mitielib/include/mitie/word_morphology_feature_extractor.h]: Why is this here?

File: mitielib/include/mitie/word_morphology_feature_extractor.h [hunk 5]
```diff
@@ -93,6 +94,7 @@ namespace mitie
                   iterator range.
         !*/
         {
+            std::vector<dlib::uint16> hits;
```
Comment (davisking) [inline mitielib/include/mitie/word_morphology_feature_extractor.h]: Why is this here?  

This is the third time I've seen variables like this added that just alias existing variables.  Please don't do this, unless there is some very good reason to do it (which is not apparent to me).

File: mitielib/src/text_categorizer.cpp [hunk 6]
```diff
@@ -71,13 +71,23 @@ namespace mitie
...
+        predict(sentence, text_tag, text_score, fe);
+    }
+
+    void text_categorizer::
+    predict (
+        const std::vector<std::string>& sentence,
+        string& text_tag,
+        double& text_score,
+        const total_word_feature_extractor& fe_
+    ) const {
```
Comment (davisking) [inline mitielib/src/text_categorizer.cpp]: Try to use the same formatting as the rest of the code, i.e. put { on it's own line.

File: mitielib/src/text_categorizer.cpp [hunk 7]
```diff
@@ -71,13 +71,23 @@ namespace mitie
...
 
         std::pair<unsigned long, double> temp;
 
-        if (fe.get_num_dimensions() == 0) {
+        if (fe_.get_num_dimensions() == 0) {
             temp = df.predict(extract_BoW_features(sentence));
         } else {
-            const std::vector<matrix<float, 0, 1> > &sent = sentence_to_feats(fe, sentence);
+            const std::vector<matrix<float, 0, 1> > &sent = sentence_to_feats(fe_, sentence);
```
Comment (davisking) [inline mitielib/src/text_categorizer.cpp]: Why is fe being renamed to fe_?

**Comment by vinvinod** [comment]
Hi Davis,

Added the required documentation for the functions and examples. 
Some redundant arguments have been removed as you have pointed out.
Also, I have removed the unnecessary temp variables I was using at some places.

**Comment by davisking** [comment]
Super, I'll look it over in a few days.

File: examples/python/text_categorizers_with_shared_feature_extractor.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+    # Load the two pure-model categorizers
+    cat1_new = text_categorizer("new_text_categorizer_1_pure_model.dat")
+    cat2_new = text_categorizer("new_text_categorizer_2_pure_model.dat")
+
+    # Now use the feature extractor to categorize the text.
+    # Observe that we are passing the same feature extractor to both the categorizers
+
+    text1 = "I am so happy"
+    pred, conf = cat1_new(tokenize(text1), text_feature_extractor)
```
Comment (davisking) [inline examples/python/text_categorizers_with_shared_feature_extractor.py]: I've looked over the PR and everything looks great, except for one more minor detail.  I noticed that if I call this without giving a feature extractor the code executes and silently does the wrong thing.  Can you make it throw an error if the user tries to make a prediction but forgets to supply the feature extractor?

Comment (vinvinod) [inline examples/python/text_categorizers_with_shared_feature_extractor.py]: I have a doubt here.
There is an option just to use the BoW if the feature extractor is not supplied. 
```
if (fe.get_num_dimensions() == 0) {
    temp = df.predict(extract_BoW_features(sentence));
} else {
    const std::vector<matrix<float, 0, 1> > &sent = sentence_to_feats(fe, sentence);
    temp = df.predict(extract_combined_features(sentence, sent));
}
```
So should it throw an error, or just give a warning that word-feature is not being used and just BoW is being used instead?

Comment (davisking) [inline examples/python/text_categorizers_with_shared_feature_extractor.py]: The user should be required to use any model in the way it was trained.  If it was trained with a particular feature set then that feature set is definitely required when it is used or the results will be very inaccurate.

Comment (vinvinod) [inline examples/python/text_categorizers_with_shared_feature_extractor.py]: We will have to store the fingerprint of the feature extractor when we are serializing the pure model. I will make the required changes and commit soon.

It won't be backward compatible! As the old pure model files wont be having the fingerprint of the the feature extractor.

**Comment by davisking** [comment]
Yeah, I should have paid more attention to this when the first PR about
pure models was submitted. :(

Anyway, what you can do is add a different version string, the first string
serialized, and use that to identify pure models that have a fingerprint.
This way you could preserve backward compatibility which is pretty
important.

**Comment by vinvinod** [comment]
I have made the required changes.
Now there is a version info saved along with the pure model file. We can increment this if there is a change in serialisation process in the future.

**Comment by davisking** [comment]
Awesome.  Thanks for the PR :)

**Comment by vinvinod** [comment]
I think you can close https://github.com/mit-nlp/MITIE/issues/36 now
